mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -243,7 +186,163 @@ FlatSymbolRefAttr ZneLowering::getOrInsertFoldedCircuit(Location loc, PatternRew
     // Remove device
     rewriter.create<quantum::DeviceReleaseOp>(loc);
     rewriter.create<func::ReturnOp>(loc, funcFolded);
-    return SymbolRefAttr::get(ctx, fnFoldedName);
+    return SymbolRefAttr::get(rewriter.getContext(), fnFoldedName);
+}
+// In *.cpp module only, to keep extraneous headers out of *.hpp
+FlatSymbolRefAttr randomLocalFolding(Location loc, PatternRewriter &rewriter,
+                                     std::string fnFoldedName, StringAttr lib, StringAttr name,
+                                     StringAttr kwargs, Type qregType, FunctionType fnFoldedType,
+                                     SmallVector<Type> typesFolded, func::FuncOp fnFoldedOp,
+                                     func::FuncOp fnAllocOp, Value numberQubitsValue,
+                                     func::FuncOp fnWithMeasurementsOp, Value c0, Value c1)
+{
+    // TODO: Implement.
+
+    // Can't throw, because disabled by compilation.
+    // throw std::logic_error("Random local folding not implemented!");
+
+    return FlatSymbolRefAttr();
+}
+// In *.cpp module only, to keep extraneous headers out of *.hpp
+FlatSymbolRefAttr allLocalFolding(Location loc, PatternRewriter &rewriter, std::string fnFoldedName,
+                                  StringAttr lib, StringAttr name, StringAttr kwargs, Type qregType,
+                                  FunctionType fnFoldedType, SmallVector<Type> typesFolded,
+                                  func::FuncOp fnFoldedOp, func::FuncOp fnAllocOp,
+                                  Value numberQubitsValue, func::FuncOp fnWithMeasurementsOp,
+                                  Value c0, Value c1)
+{
+    // Allocate qubits
+    Value allocQreg = rewriter.create<func::CallOp>(loc, fnAllocOp, numberQubitsValue).getResult(0);
+
+    int64_t sizeArgs = fnFoldedOp.getArguments().size();
+    Value size = fnFoldedOp.getArgument(sizeArgs - 1);
+
+    // Walk through the operations in fnWithMeasurementsOp
+    fnWithMeasurementsOp.walk([&](quantum::QuantumGate op) {
+        // Insert a for loop immediately before each quantum::QuantumGate
+        auto innerLoc = op->getLoc();
+        rewriter.setInsertionPoint(op);
+        rewriter.create<scf::ForOp>(
+            innerLoc, c0, size, c1, ValueRange(),
+            [&](OpBuilder &builder, Location forLoc, Value i, ValueRange iterArgs) {
+                // Set insertion point within the loop
+                builder.setInsertionPointToEnd(builder.getBlock());
+                // Create adjoint and original operations
+                auto origOp = builder.clone(*op)->getResult(0);
```

Comment:
That looks on a good track, what are the remaining issues for this folding? It looks like you might need to extract the register and give it to the forloop, in order to glue everything together.